lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1265,10 +1379,117 @@ void DShrPrimOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.insert<patterns::DShrOfConstant>(context);
 }
 
+namespace {
+
+/// Canonicalize a cat tree into single cat operation.
+class FlattenCat : public mlir::RewritePattern {
```

Comment:
This seems fine.

It would be nice if there was a way that the cat rely on the recursive canonicalization to make this simpler. E.g., if this could only work at the level of inlining a cat into another cat and then this would naturally flatten the tree structure. However, I think this would require multiple explicit recursive canonicalizations as opposed to it being automatic.

Any idea if this just works with the simpler version? This can be a follow-on discussion.